enhancements/automated-service-ca-rotation.md

```diff
@@ -0,0 +1,246 @@
+  current CA to determine whether the serving cert was issued by the current CA.
+  - If the serving cert is determined not to have been issued by the current CA,
+    it needs to be regenerated.
+  - Previously, `Subject.CommonName` of a serving cert was compared with
+    `Issuer.CommonName` of the CA. Since ensuring trust between old and new CAs
+    requires that `Subject.CommonName` be reused, this comparison will no longer
+    be sufficient to determine lineage.
+  - This requires that all new certificates - whether CA or entity - have
+    `AuthorityKeyId` and `SubjectKeyId` set.
+  - In addition to ensuring regeneration when `AuthorityKeyId` and `SubjectKeyId`
+    are set, this comparison schema will also ensure regeneration of serving
+    certs without `SubjectKeyId` set (due to being generated by a CA without
+    `AuthorityKeyId` set) since an empty `SubjectKeyId` will not match a
+    populated `AuthorityKeyId`.
+
+- Provide an indication that rotation has taken place via an event.
+
+- Allow rotation to be forced when
+  `ServiceCA.Spec.ForceRotation.Reason` is set to a non-empty value
+  that does not match the value for the
+  `service.beta.openshift.io/forced-rotation-reason` annotation on the
+  signing secret.
+  - Setting an annotation on the signing secret ensures that a forced
+    rotation is triggered at most once per reason.
+  - If a forced rotation and expiry-triggered rotation coincide, only
+    the forced rotation should be performed.
+
+- Allow the duration of the grace period for continuing to trust the
+  previous CA to be overridden when forcing rotation by setting
+  `ServiceCA.Spec.ForceRotation.GraceDuration`.
+  - If set, the expiry of the intermediate certs bridging trust
+    between the pre-rotation serving certs and CA bundle will be set
+    to the time of rotation plus the provided duration.
+
+### Risks and Mitigations
+
+Automatic rotation is only effective if consumers of certs and CA bundles will
+also refresh automatically before expiry of the pre-rotation CA.
+
+The proposed rotation scheme does not extend the useful lifespan of serving certs
+and CA bundles from the pre-rotation CA. Their lifespan will still be 1
+year. Instead, the rotation schema provides a 6 month window when both the old
+and new CAs will be trusted. Within that 6 month window, it is critical that
+services that consume serving certs and clients that consume CA bundles start
+using the serving certs and CA bundle provided by the new CA. For some consumers
+this may happen automatically (e.g. via restart on change or hitless
+rotation. Others may need to be manually restarted.
+
+OpenShift 4.1 Beta 5 - the first 4.1 beta candidate provided to external
+audiences - was made available on May 14, 2019. Without automatic or manual
+rotation, the control plane of a cluster installed on that date would start
+failing a year later - May 14, 2020 - due to the expiry of the service CA
+generated at the time of cluster deployment.
```

Comment:
@knewcomerRH fyi ^^ those dates

Reply:
thanks @sttts Great to see this working being tackled.